apypie/action.py

```diff
@@ -217,17 +217,19 @@ def prepare_params(self, input_dict):
         params.update(route_params)
         return params
 
-    def _prepare_params(self, action_params, input_dict):
-        # type: (Iterable[Param], dict) -> dict
+    def _prepare_params(self, action_params, input_dict, input_subdict=None):
+        # type: (Iterable[Param], dict, Optional[dict]) -> dict
         result = {}
 
         for param in action_params:
             if param.expected_type == 'hash' and param.params:
-                nested_result = self._prepare_params(param.params, input_dict)
+                nested_result = self._prepare_params(param.params, input_dict, input_dict.get(param.name))
                 if nested_result:
                     result[param.name] = nested_result
             elif param.name in input_dict:
                 result[param.name] = input_dict[param.name]
+            elif input_subdict and param.name in input_subdict:
+                result[param.name] = input_subdict[param.name]
```

Comment:
Not sure, I fully understand it.
1. Wouldn't it make more sense to do the recursive call with a merged version of those dicts? Then you could apply the mechanism to any finite depth.
2. I'd suspect that the value from the subdict is more specific by nature. Should it not take precedence?

Reply:
you're suggesting something like:

```python
          for param in action_params:
              if param.expected_type == 'hash' and param.params:
                  nested_input = input_dict.copy()
                  nested_input.update(input_dict.get(param.name, {}))
                  nested_result = self._prepare_params(param.params, nested_input)
                  if nested_result:
                      result[param.name] = nested_result
              elif param.name in input_dict:
                  result[param.name] = input_dict[param.name]
```

Reply:
this does pass the test below, but my fear would be that the sub-dict might have a key that's also in the outer dict (like `id`) which has a different value and would thus override it?

Reply:
huh, this also works, and is simpler?

```python
          for param in action_params:
              if param.expected_type == 'hash' and param.params:
                  nested_dict = input_dict.get(param.name, input_dict)
                  nested_result = self._prepare_params(param.params, nested_dict)
                  if nested_result:
                      result[param.name] = nested_result
              elif param.name in input_dict:
                  result[param.name] = input_dict[param.name]
```

Reply:
it'd obviously break if `input_dict[param.name]` is not actually a dict, but why wouldn't it? :)

Reply:
I thought you merged them because it might still want items from the outer dict.
And yes the overwriting is the reason i said, you should probably prioritize the subdict.

Reply:
> it'd obviously break if `input_dict[param.name]` is not actually a dict, but why wouldn't it? :)

In that case it'd break anyway.

Reply:
I don't think we should end up mixing inner and outer dicts, so I prefer my last proposal "take the inner if there is any, otherwise the outer"